File: Sources/Services/PlayerProfile.swift
import Foundation
import Observation

@MainActor
@Observable
final class PlayerProfile {
    private enum Key {
        static let tutorial = "collapse.tutorial.completed"
        static let legacyBestScore = "collapse.best.score"
        static let skin = "collapse.skin.selected"
        static let dailyStreak = "collapse.daily.streak"
        static let lastRunDate = "collapse.daily.lastRun"
        static let legacyDailyBest = "collapse.daily.bestScore"
        static let legacyTopScores = "collapse.local.topScores"
        static let gemBalance = "collapse.economy.gems"
        static let unlockedSkins = "collapse.skin.unlocked"
        static let mode = "collapse.mode.selected"
        static let language = "collapse.language.selected"
        static let modeBest = "collapse.mode.best"
        static let modeDailyBest = "collapse.mode.dailyBest"
        static let modeTopScores = "collapse.mode.topScores"
        static let musicEnabled = "collapse.audio.music"
        static let soundEnabled = "collapse.audio.sound"
        static let hapticsEnabled = "collapse.audio.haptics"
    }

    private let defaults: UserDefaults
    private var modeBestScores: [String: Int]
    private var modeDailyBestScores: [String: Int]
    private var modeTopScores: [String: [Int]]

    var didCompleteTutorial: Bool {
        didSet { defaults.set(didCompleteTutorial, forKey: Key.tutorial) }
    }

    var selectedSkin: GameSkin {
        didSet { defaults.set(selectedSkin.rawValue, forKey: Key.skin) }
    }

    var selectedMode: GameMode {
        didSet { defaults.set(selectedMode.rawValue, forKey: Key.mode) }
    }

    var selectedLanguage: AppLanguage {
        didSet { defaults.set(selectedLanguage.rawValue, forKey: Key.language) }
    }

    var musicEnabled: Bool {
        didSet { defaults.set(musicEnabled, forKey: Key.musicEnabled) }
    }

    var soundEnabled: Bool {
        didSet { defaults.set(soundEnabled, forKey: Key.soundEnabled) }
    }

    var hapticsEnabled: Bool {
        didSet { defaults.set(hapticsEnabled, forKey: Key.hapticsEnabled) }
    }

    private(set) var dailyRunStreak: Int {
        didSet { defaults.set(dailyRunStreak, forKey: Key.dailyStreak) }
    }

    private(set) var lastRunDate: Date? {
        didSet { defaults.set(lastRunDate, forKey: Key.lastRunDate) }
    }

    private(set) var gemBalance: Int {
        didSet { defaults.set(gemBalance, forKey: Key.gemBalance) }
    }

    private(set) var unlockedSkinIDs: Set<String> {
        didSet { defaults.set(unlockedSkinIDs.sorted(), forKey: Key.unlockedSkins) }
    }

    var bestScore: Int { bestScore(for: selectedMode) }
    var dailyBestScore: Int { dailyBestScore(for: selectedMode) }
    var topScores: [Int] { topScores(for: selectedMode) }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        didCompleteTutorial = defaults.bool(forKey: Key.tutorial)
        selectedSkin = GameSkin(rawValue: defaults.string(forKey: Key.skin) ?? "") ?? .classic
        selectedMode = GameMode(rawValue: defaults.string(forKey: Key.mode) ?? "") ?? .classic
        selectedLanguage = AppLanguage(rawValue: defaults.string(forKey: Key.language) ?? "") ?? .systemDefault
        musicEnabled = defaults.object(forKey: Key.musicEnabled) as? Bool ?? true
        soundEnabled = defaults.object(forKey: Key.soundEnabled) as? Bool ?? true
        hapticsEnabled = defaults.object(forKey: Key.hapticsEnabled) as? Bool ?? true
        dailyRunStreak = defaults.integer(forKey: Key.dailyStreak)
        lastRunDate = defaults.object(forKey: Key.lastRunDate) as? Date
        gemBalance = defaults.integer(forKey: Key.gemBalance)
        unlockedSkinIDs = Set(defaults.stringArray(forKey: Key.unlockedSkins) ?? [])
        modeBestScores = Self.readIntDictionary(defaults, key: Key.modeBest)
        modeDailyBestScores = Self.readIntDictionary(defaults, key: Key.modeDailyBest)
        modeTopScores = Self.readScoreDictionary(defaults, key: Key.modeTopScores)
        migrateLegacyStatsIfNeeded()
    }

    func registerRunStart(now: Date = .now, calendar: Calendar = .current) {
        let today = calendar.startOfDay(for: now)
        guard let lastRunDate else {
            dailyRunStreak = 1
            modeDailyBestScores = [:]
            persistModeDailyBest()
            self.lastRunDate = today
            return
        }

        let previousDay = calendar.startOfDay(for: lastRunDate)
        guard previousDay != today else { return }
        let yesterday = calendar.date(byAdding: .day, value: -1, to: today)
        dailyRunStreak = previousDay == yesterday ? max(1, dailyRunStreak + 1) : 1
        modeDailyBestScores = [:]
        persistModeDailyBest()
        self.lastRunDate = today
    }

    func record(score: Int, gemsEarned: Int, mode: GameMode) async {
        guard mode.isCompetitive else { return }
        let key = mode.rawValue
        modeBestScores[key] = max(modeBestScores[key] ?? 0, score)
        modeDailyBestScores[key] = max(modeDailyBestScores[key] ?? 0, score)
        gemBalance += max(0, gemsEarned)
        if score > 0 {
            modeTopScores[key] = Array(((modeTopScores[key] ?? []) + [score]).sorted(by: >).prefix(3))
        }
        persistModeStats()
    }

    func bestScore(for mode: GameMode) -> Int {
        modeBestScores[mode.rawValue] ?? 0
    }

    func dailyBestScore(for mode: GameMode) -> Int {
        modeDailyBestScores[mode.rawValue] ?? 0
    }

    func topScores(for mode: GameMode) -> [Int] {
        modeTopScores[mode.rawValue] ?? []
    }

    func localRank(for score: Int, mode: GameMode? = nil) -> Int? {
        let targetMode = mode ?? selectedMode
        guard score > 0 else { return nil }
        let projected = (topScores(for: targetMode) + [score]).sorted(by: >)
        return projected.firstIndex(of: score).map { $0 + 1 }
    }

    func isUnlocked(_ skin: GameSkin, isPlusUnlocked: Bool) -> Bool {
        switch skin.access {
        case .free:
            true
        case .gems:
            unlockedSkinIDs.contains(skin.id)
        case .plus:
            isPlusUnlocked
        }
    }

    func unlock(_ skin: GameSkin) -> Bool {
        guard case .gems(let cost) = skin.access else { return false }
        guard !unlockedSkinIDs.contains(skin.id) else { return true }
        guard gemBalance >= cost else { return false }
        gemBalance -= cost
        unlockedSkinIDs.insert(skin.id)
        return true
    }

    func activeSkin(isPlusUnlocked: Bool) -> GameSkin {
        guard isUnlocked(selectedSkin, isPlusUnlocked: isPlusUnlocked) else { return .classic }
        return selectedSkin
    }

    private func migrateLegacyStatsIfNeeded() {
        let key = GameMode.classic.rawValue
        if modeBestScores.isEmpty {
            let legacy = defaults.integer(forKey: Key.legacyBestScore)
            if legacy > 0 { modeBestScores[key] = legacy }
        }
        if modeDailyBestScores.isEmpty {
            let legacy = defaults.integer(forKey: Key.legacyDailyBest)
            if legacy > 0 { modeDailyBestScores[key] = legacy }
        }
        if modeTopScores.isEmpty {
            let legacy = (defaults.array(forKey: Key.legacyTopScores) as? [Int] ?? []).sorted(by: >)
            if !legacy.isEmpty { modeTopScores[key] = Array(legacy.prefix(3)) }
        }
        persistModeStats()
    }

    private func persistModeStats() {
        defaults.set(modeBestScores, forKey: Key.modeBest)
        persistModeDailyBest()
        defaults.set(modeTopScores, forKey: Key.modeTopScores)
    }

    private func persistModeDailyBest() {
        defaults.set(modeDailyBestScores, forKey: Key.modeDailyBest)
    }

    private static func readIntDictionary(_ defaults: UserDefaults, key: String) -> [String: Int] {
        guard let raw = defaults.dictionary(forKey: key) else { return [:] }
        return raw.compactMapValues { ($0 as? NSNumber)?.intValue }
    }

    private static func readScoreDictionary(_ defaults: UserDefaults, key: String) -> [String: [Int]] {
        guard let raw = defaults.dictionary(forKey: key) else { return [:] }
        return raw.reduce(into: [:]) { result, pair in
            guard let values = pair.value as? [Int] else { return }
            result[pair.key] = Array(values.sorted(by: >).prefix(3))
        }
    }
}

File: android/app/src/main/java/com/collapse/game/services/PlayerProfile.kt
package com.collapse.game.services

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.collapse.game.domain.GameMode
import com.collapse.game.ui.GameSkin
import com.collapse.game.ui.SkinAccess
import java.time.LocalDate

class PlayerProfile(context: Context) {
    private val preferences = context.getSharedPreferences("collapse.profile", Context.MODE_PRIVATE)

    var didCompleteTutorial by mutableStateOf(preferences.getBoolean(KEY_TUTORIAL, false))
        private set
    var dailyRunStreak by mutableIntStateOf(preferences.getInt(KEY_STREAK, 0))
        private set
    var gemBalance by mutableIntStateOf(preferences.getInt(KEY_GEMS, 0))
        private set
    var selectedSkin by mutableStateOf(readSelectedSkin())
        private set
    var selectedMode by mutableStateOf(readSelectedMode())
        private set
    var unlockedSkinIds by mutableStateOf(preferences.getStringSet(KEY_UNLOCKED, emptySet())?.toSet() ?: emptySet())
        private set
    var musicEnabled by mutableStateOf(preferences.getBoolean(KEY_MUSIC, true))
        private set
    var soundEnabled by mutableStateOf(preferences.getBoolean(KEY_SOUND, true))
        private set
    var hapticsEnabled by mutableStateOf(preferences.getBoolean(KEY_HAPTICS, true))
        private set

    val bestScore: Int
        get() = bestScore(selectedMode)
    val dailyBestScore: Int
        get() = dailyBestScore(selectedMode)

    init {
        migrateLegacyStatsIfNeeded()
    }

    fun completeTutorial() {
        didCompleteTutorial = true
        preferences.edit().putBoolean(KEY_TUTORIAL, true).apply()
    }

    fun selectMode(mode: GameMode) {
        selectedMode = mode
        preferences.edit().putString(KEY_MODE, mode.name).apply()
    }

    fun updateMusicEnabled(enabled: Boolean) {
        musicEnabled = enabled
        preferences.edit().putBoolean(KEY_MUSIC, enabled).apply()
    }

    fun updateSoundEnabled(enabled: Boolean) {
        soundEnabled = enabled
        preferences.edit().putBoolean(KEY_SOUND, enabled).apply()
    }

    fun updateHapticsEnabled(enabled: Boolean) {
        hapticsEnabled = enabled
        preferences.edit().putBoolean(KEY_HAPTICS, enabled).apply()
    }

    fun registerRunStart(today: LocalDate = LocalDate.now()) {
        val last = preferences.getString(KEY_LAST_RUN_DATE, null)?.let(LocalDate::parse)
        if (last == today) return
        dailyRunStreak = if (last == today.minusDays(1)) maxOf(1, dailyRunStreak + 1) else 1
        clearDailyModeScores()
        preferences.edit()
            .putInt(KEY_STREAK, dailyRunStreak)
            .putString(KEY_LAST_RUN_DATE, today.toString())
            .apply()
    }

    fun record(score: Int, gemsEarned: Int, mode: GameMode) {
        if (!mode.isCompetitive) return
        val best = maxOf(bestScore(mode), score)
        val daily = maxOf(dailyBestScore(mode), score)
        val top = if (score > 0) (topScores(mode) + score).sortedDescending().take(3) else topScores(mode)
        gemBalance += maxOf(0, gemsEarned)
        preferences.edit()
            .putInt(bestKey(mode), best)
            .putInt(dailyBestKey(mode), daily)
            .putString(topScoresKey(mode), top.joinToString(","))
            .putInt(KEY_GEMS, gemBalance)
            .apply()
    }

    fun bestScore(mode: GameMode): Int = preferences.getInt(bestKey(mode), 0)

    fun dailyBestScore(mode: GameMode): Int = preferences.getInt(dailyBestKey(mode), 0)

    fun topScores(mode: GameMode): List<Int> = preferences
        .getString(topScoresKey(mode), null)
        ?.split(',')
        ?.mapNotNull(String::toIntOrNull)
        ?.sortedDescending()
        ?.take(3)
        ?: emptyList()

    fun localRank(score: Int, mode: GameMode = selectedMode): Int? {
        if (score <= 0) return null
        val projected = (topScores(mode) + score).sortedDescending()
        return projected.indexOf(score).takeIf { it >= 0 }?.plus(1)
    }

    fun selectSkin(skin: GameSkin, isPlusUnlocked: Boolean): Boolean {
        if (!isUnlocked(skin, isPlusUnlocked)) return false
        selectedSkin = skin
        preferences.edit().putString(KEY_SKIN, skin.name).apply()
        return true
    }

    fun unlock(skin: GameSkin): Boolean {
        val access = skin.access as? SkinAccess.Gems ?: return false
        if (unlockedSkinIds.contains(skin.name)) return true
        if (gemBalance < access.cost) return false
        gemBalance -= access.cost
        unlockedSkinIds = unlockedSkinIds + skin.name
        preferences.edit()
            .putInt(KEY_GEMS, gemBalance)
            .putStringSet(KEY_UNLOCKED, unlockedSkinIds)
            .apply()
        return true
    }

    fun isUnlocked(skin: GameSkin, isPlusUnlocked: Boolean): Boolean = when (skin.access) {
        SkinAccess.Free -> true
        is SkinAccess.Gems -> unlockedSkinIds.contains(skin.name)
        SkinAccess.Plus -> isPlusUnlocked
    }

    fun activeSkin(isPlusUnlocked: Boolean): GameSkin =
        if (isUnlocked(selectedSkin, isPlusUnlocked)) selectedSkin else GameSkin.Classic

    private fun migrateLegacyStatsIfNeeded() {
        val classic = GameMode.Classic
        val editor = preferences.edit()
        if (!preferences.contains(bestKey(classic))) {
            editor.putInt(bestKey(classic), preferences.getInt(KEY_LEGACY_BEST, 0))
        }
        if (!preferences.contains(dailyBestKey(classic))) {
            editor.putInt(dailyBestKey(classic), preferences.getInt(KEY_LEGACY_DAILY_BEST, 0))
        }
        editor.apply()
    }

    private fun clearDailyModeScores() {
        val editor = preferences.edit()
        GameMode.entries.forEach { editor.remove(dailyBestKey(it)) }
        editor.apply()
    }

    private fun bestKey(mode: GameMode) = "mode.best.${mode.name}"
    private fun dailyBestKey(mode: GameMode) = "mode.daily.${mode.name}"
    private fun topScoresKey(mode: GameMode) = "mode.top.${mode.name}"

    private fun readSelectedSkin(): GameSkin {
        val raw = preferences.getString(KEY_SKIN, null) ?: return GameSkin.Classic
        return GameSkin.entries.firstOrNull { it.name == raw } ?: GameSkin.Classic
    }

    private fun readSelectedMode(): GameMode {
        val raw = preferences.getString(KEY_MODE, null) ?: return GameMode.Classic
        return GameMode.entries.firstOrNull { it.name == raw } ?: GameMode.Classic
    }

    private companion object {
        const val KEY_TUTORIAL = "tutorial.completed"
        const val KEY_LEGACY_BEST = "best.score"
        const val KEY_LEGACY_DAILY_BEST = "daily.best"
        const val KEY_STREAK = "daily.streak"
        const val KEY_LAST_RUN_DATE = "daily.lastRun"
        const val KEY_GEMS = "economy.gems"
        const val KEY_SKIN = "skin.selected"
        const val KEY_UNLOCKED = "skin.unlocked"
        const val KEY_MODE = "mode.selected"
        const val KEY_MUSIC = "audio.music"
        const val KEY_SOUND = "audio.sound"
        const val KEY_HAPTICS = "audio.haptics"
    }
}
